easybuild/framework/easyconfig/tools.py

```diff
@@ -333,3 +348,100 @@ def stats_to_str(stats):
         txt += "%s%s: %s,\n" % (pref, quote_str(k), quote_str(v))
     txt += "}"
     return txt
+
+
+def find_related_easyconfigs(path, ec):
+    """
+    Find related easyconfigs for provided parsed easyconfig in specified path.
+
+    A list of easyconfigs for the same software (name) is returned,
+    matching the 1st criterion that yields a non-empty list.
+
+    Software version is considered more important than toolchain.
+
+    Toolchain is considered with exact same version prior to without version (only name).
+
+    Matching versionsuffix is considered prior to any versionsuffix.
+
+    Exact software version is considered prior to matching major/minor version numbers,
+    and only matching major version number. Any software version is considered last.
+
+    The following criteria are considered, in order (with 'version criterion' being either an
+    exact version match, a major/minor version match, a major version match, or no version match).
+
+    (i)   software version criterion, matching versionsuffix and toolchain name/version
+    (ii)  software version criterion, matching versionsuffix and toolchain name (any toolchain version)
+    (iii) software version criterion, matching versionsuffix (any toolchain name/version)
+    (iv)  software version criterion, matching toolchain name/version (any versionsuffix)
+    (v)   software version criterion, matching toolchain name (any versionsuffix, toolchain version)
+    (vi)  software version criterion (any versionsuffix, toolchain name/version)
+
+    If no related easyconfigs with a matching software name are found, an empty list is returned.
+    """
+    name = ec.name
+    version = ec.version
+    versionsuffix = ec['versionsuffix']
+    toolchain_name = ec['toolchain']['name']
+    toolchain_name_pattern = r'-%s-\S+' % toolchain_name
+    toolchain = "%s-%s" % (toolchain_name, ec['toolchain']['version'])
+    toolchain_pattern = '-%s' % toolchain
+    if toolchain_name == DUMMY_TOOLCHAIN_NAME:
+        toolchain_name_pattern = ''
+        toolchain_pattern = ''
+
+    potential_paths = [glob.glob(ec_path) for ec_path in create_paths(path, name, '*')]
+    potential_paths = sum(potential_paths, [])  # flatten
+
+    parsed_version = LooseVersion(version).version
+    version_patterns = [version]  # exact version match
+    if len(parsed_version) >= 2:
+        version_patterns.append(r'%s\.%s\.[0-9_A-Za-z]+' % tuple(parsed_version[:2]))  # major/minor version match
+    if parsed_version != parsed_version[0]:
+        version_patterns.append(r'%s\.[0-9-]+\.[0-9_A-Za-z]+' % parsed_version[0])  # major version match
+    version_patterns.append(r'[0-9._A-Za-z-]+')  # any version
+
+    regexes = []
+    for version_pattern in version_patterns:
+        regexes.extend([
+            re.compile((r"^\S+/%s-%s%s%s.eb$" % (name, version_pattern, toolchain_pattern, versionsuffix))),
```

Comment:
do the `re.compile` in a separate step on the whole list of patterns (use `map`).
make a pattern that holds all the common pattern `r"^\S+/%s-%s%%s.eb$" %(name, version_pattern)` is a good start